modules/graph/fragment/arrow_fragment.h

```diff
@@ -629,20 +630,440 @@ class ArrowFragment
+      nbytes += vec[i]->nbytes();                                          \
+    }                                                                      \
+  } while (0);
+
+#define GENERATE_VEC_VEC_META(prefix, vec, v_label_num, e_label_num,          \
+                              start_v_num, start_e_num)                       \
+  do {                                                                        \
+    for (label_id_t i = 0; i < v_label_num; ++i) {                            \
+      for (label_id_t j = 0; j < e_label_num; ++j) {                          \
+        new_meta.AddMember(generate_name_with_suffix(prefix, i + start_v_num, \
+                                                     j + start_e_num),        \
+                           vec[i][j]->meta());                                \
+        nbytes += vec[i][j]->nbytes();                                        \
+      }                                                                       \
+    }                                                                         \
+  } while (0);
+
+  boost::leaf::result<ObjectID> AddEdge(
+      Client& client, std::vector<std::shared_ptr<arrow::Table>>& edge_tables,
+      const std::vector<std::set<std::pair<std::string, std::string>>>&
+          edge_relations,
+      int concurrency) {
+    int extra_edge_label_num = edge_tables.size();
+    int total_edge_label_num = edge_label_num_ + extra_edge_label_num;
+
+    // Newly constructed data structures
+    vineyard::Array<vid_t> vy_ovnums, vy_tvnums;
+    std::vector<std::shared_ptr<vineyard::NumericArray<vid_t>>> vy_ovgid_lists;
+    std::vector<std::shared_ptr<vineyard::Hashmap<vid_t, vid_t>>> vy_ovg2l_maps;
+    std::vector<std::shared_ptr<vineyard::Table>> vy_edge_tables;
+
+    std::vector<std::vector<std::shared_ptr<vineyard::FixedSizeBinaryArray>>>
+        vy_ie_lists, vy_oe_lists;
+    std::vector<std::vector<std::shared_ptr<vineyard::NumericArray<int64_t>>>>
+        vy_ie_offsets_lists, vy_oe_offsets_lists;
+
+    // Init size
+    vy_ovgid_lists.resize(vertex_label_num_);
+    vy_ovg2l_maps.resize(vertex_label_num_);
+    vy_edge_tables.resize(extra_edge_label_num);
+
+    if (directed_) {
+      vy_ie_lists.resize(vertex_label_num_);
+      vy_ie_offsets_lists.resize(vertex_label_num_);
+    }
+    vy_oe_lists.resize(vertex_label_num_);
+    vy_oe_offsets_lists.resize(vertex_label_num_);
+
+    for (label_id_t i = 0; i < vertex_label_num_; ++i) {
+      if (directed_) {
+        vy_ie_lists[i].resize(extra_edge_label_num);
+        vy_ie_offsets_lists[i].resize(extra_edge_label_num);
+      }
+      vy_oe_lists[i].resize(extra_edge_label_num);
+      vy_oe_offsets_lists[i].resize(extra_edge_label_num);
+    }
+
+    // Make a copy of ovg2l map, since we need to add some extra outer vertices
+    // pulled in this fragment by new edges.
+    std::vector<ska::flat_hash_map<vid_t, vid_t>> ovg2l_maps(vertex_label_num_);
```

Comment:
Use the hashmap builder rather than `ska::flat_hash_map<vid_t, vid_t>`.

Reply:
I reused the function `generate_outer_vertices_map` in line 768.